File: src/main/scala/jk_5/nailed/server/map/gamerule/WrappedEditableGameRules.scala
package jk_5.nailed.server.map.gamerule

import jk_5.nailed.api.mappack.gamerule.{EditableGameRules, GameRules}

import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
case class WrappedEditableGameRules(gameRules: GameRules) extends EditableGameRules {

  val rules = mutable.HashMap[String, String]()

  for(rule <- gameRules.list){
    rules.put(rule, gameRules(rule))
  }

  override def apply(key: String): String = this.rules.getOrElse(key, "")
  override def update(key: String, value: String): Unit = this.rules.put(key, value)
  override def ruleExists(key: String) = this.rules.contains(key)
  override def list: Seq[String] = this.rules.keySet.toSeq

  override def toString = s"WrappedEditableGameRules{wrapped=$gameRules}"
}

File: src/main/java/jk_5/nailed/server/map/gamerule/WrappedEditableGameRules.java
package jk_5.nailed.server.map.gamerule;

import jk_5.nailed.api.gamerule.*;

import javax.annotation.Nonnull;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * No description given
 *
 * @author jk-5
 */
public class WrappedEditableGameRules implements EditableGameRules {

    private final Map<GameRuleKey<?>, EditableGameRule<?>> gameRuleMap = new HashMap<GameRuleKey<?>, EditableGameRule<?>>();

    public WrappedEditableGameRules(GameRules wrapped) {
        //noinspection unchecked
        for(GameRule<Object> rule : (Collection<GameRule<Object>>) wrapped.list()){
            EditableGameRule<Object> newRule = new DefaultEditableGameRule<Object>(rule.getKey());
            newRule.setValue(rule.getValue());
            gameRuleMap.put(rule.getKey(), newRule);
        }
    }

    @Nonnull
    @Override
    public <T> EditableGameRule<T> get(@Nonnull GameRuleKey<T> key) {
        //noinspection unchecked
        EditableGameRule<T> rule = (EditableGameRule<T>) gameRuleMap.get(key);
        if(rule == null){
            rule = new DefaultEditableGameRule<T>(key);
        }
        return rule;
    }

    @Nonnull
    @Override
    public Collection<EditableGameRule<?>> list() {
        return Collections.unmodifiableCollection(gameRuleMap.values());
    }

    @Override
    public String toString() {
        return "WrappedEditableGameRules{" +
                "gameRuleMap=" + gameRuleMap +
                '}';
    }
}
